[android][test] Fix a handful of tests and disable one CxxToSwiftToCxx bridging test

finagolfin opened the pull request:
Also, make the analogous change to https://github.com/apple/swift-driver/pull/1372, which gets the sanitizer tests working on Android again, and remove the `lld_lto` feature in the tests, which is now unused.

[The first test initially worked](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android/3472/consoleText), but then something changed in trunk to add the `noundef`, breaking the test.

The second broke a couple weeks ago because of the last command line added in #65662, as I think those C++ modules don't currently work on Android, so I disabled the test.

@hyp and @drodriguez, this should get the Android CI passing again.

finagolfin committed:
[android][test] Fix a handful of tests and disable one CxxToSwiftToCxx bridging test

Also, make the analogous change to apple/swift-driver#1372, which gets the
sanitizer tests working on Android again, and remove the lld_lto feature in the
tests, which is now unused.

drodriguez commented:
@swift-ci please test

drodriguez commented:
I agree with disabling the second test.

Not sure about the `noundef`, since the check will allow it in every platform, and it might be incorrect for those other platforms. I leave that to decide to others that might have more context.

finagolfin commented:
@hyp, any idea why the `noundef` is only added for Android armv7, breaking this first test there?

finagolfin commented:
@egorzhdan, maybe you can review?

finagolfin commented:
Rebased and added fix for just-added IRGen/section.swift test from #65901 that is [failing on the Android CI](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android-arm64/4149/consoleText):
```
       67:  .type $s7section2g0_Wz,@object 
       68:  .local $s7section2g0_Wz 
       69:  .comm $s7section2g0_Wz,8,8 
       70:  .hidden $s7section2g0Sivp 
       71:  .type $s7section2g0Sivp,@object 
       72:  .section "__TEXT,__mysection","aw",@progbits,unique,1 
not:41      !~~~~~~~                                               error: no match expected
       73:  .globl $s7section2g0Sivp 
       74:  .p2align 3 
       75: $s7section2g0Sivp: 
       76:  .xword 1 
       77:  .size $s7section2g0Sivp, 8
```
@kubamracek, I haven't tested this fix, but am just following your lead on the syntax and the test output shown above. Let me know if this should work.

finagolfin commented:
@drodriguez, can you approve and merge? We've given the C++ Interop people enough time to chime in, meanwhile this will get the Android CI green again.

kubamracek commented on test/IRGen/section.swift:
```diff
@@ -38,6 +38,8 @@
 // ASM: $s7section3fooyyF:
 // ASM-linux-gnu: .section{{.*}}__TEXT,__mysection
+// ASM-linux-android: .section{{.*}}__TEXT,__mysection
+// ASM-linux-androideabi: .section{{.*}}__TEXT,__mysection
```
This looks great.

finagolfin replied:
I have now tested this and made sure it works.

drodriguez commented on test/Interop/Cxx/class/copy-move-assignment-irgen.swift:
```diff
@@ -20,7 +20,7 @@ public func copyAssign() {
 // CHECK: call {{void|\%struct.NonTrivialCopyAndCopyMoveAssign\*}} @{{_ZN31NonTrivialCopyAndCopyMoveAssignC1Ev|_ZN31NonTrivialCopyAndCopyMoveAssignC2Ev|"\?\?0NonTrivialCopyAndCopyMoveAssign@@QEAA@XZ"}}(%struct.NonTrivialCopyAndCopyMoveAssign* %[[COPY_INSTANCE2:.*]])
 // CHECK: call {{void|\%struct.NonTrivialCopyAndCopyMoveAssign\*}} @{{_ZN31NonTrivialCopyAndCopyMoveAssignD1Ev|_ZN31NonTrivialCopyAndCopyMoveAssignD2Ev|"\?\?1NonTrivialCopyAndCopyMoveAssign@@QEAA@XZ"}}(%struct.NonTrivialCopyAndCopyMoveAssign* %[[COPY_INSTANCE]])
-// CHECK: call {{void|\%struct.NonTrivialCopyAndCopyMoveAssign\*}} @{{_ZN31NonTrivialCopyAndCopyMoveAssignC1EOS_|_ZN31NonTrivialCopyAndCopyMoveAssignC2EOS_|_ZN31NonTrivialCopyAndCopyMoveAssignC1ERKS_Tm|_ZN31NonTrivialCopyAndCopyMoveAssignC2ERKS_Tm|"\?\?0NonTrivialCopyAndCopyMoveAssign@@QEAA@AEBU0@@Z"}}(
+// CHECK: call {{(noundef )?}}{{void|\%struct.NonTrivialCopyAndCopyMoveAssign\*}} @{{_ZN31NonTrivialCopyAndCopyMoveAssignC1EOS_|_ZN31NonTrivialCopyAndCopyMoveAssignC2EOS_|_ZN31NonTrivialCopyAndCopyMoveAssignC1ERKS_Tm|_ZN31NonTrivialCopyAndCopyMoveAssignC2ERKS_Tm|"\?\?0NonTrivialCopyAndCopyMoveAssign@@QEAA@AEBU0@@Z"}}(
```
From what I can read from the LLVM IR manual, `noundef` should only apply to the `%struct`, right? Can you confirm that Android is returning the `%struct` and not `void`?

If that's the case, to reduce the affected area it should be moved to be only in front of the `%struct` and not apply to a possible `void`.

```
CHECK: call {{void|(noundef )?\%struct.NonTrivialCopyAndCopyMoveAssign\*}}
```

I am still unsure why Android needs that, and if it is safe to apply to every platform. It seems to be a stricter version of the naked type, but I am not sure. The least we can do is to apply it the type that it should be affecting.

finagolfin commented:
Rebased, updated the `noundef` test, and added a driver change.

finagolfin commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -42,8 +42,7 @@ std::string
                                                  bool shared) const {
   return (Twine("libclang_rt.") + Sanitizer + "-" +
-          this->getTriple().getArchName() +
-          (this->getTriple().isAndroid() ? "-android" : "") + ".a")
```
I had [added this several years ago](https://github.com/apple/swift/pull/27167/files#r353616826), but it is no longer needed: see apple/swift-driver#1372 for more info.

finagolfin commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1452,9 +1452,11 @@ void ToolChain::getClangLibraryPath(const ArgList &Args,
+      T.isOSDarwin() ? "darwin" : getPlatformNameForTriple(Triple);
+  if (platformName == "android")
+    platformName = "linux";
+  llvm::sys::path::append(LibPath, "clang", "lib", platformName);
```
This is a translation of apple/swift-driver#1372, see that pull for more info.

compnerd replied:
I wonder if we can use `llvm::Triple::getOSName` instead.

finagolfin replied:
> I wonder if we can use llvm::Triple::getOSName instead.

Just looked into this: [clang uses `ToolChain::getOSLibName()` instead](https://github.com/llvm/llvm-project/blob/bbc0f99f3bc96f1db16f649fc21dd18e5b0918f6/clang/lib/Driver/ToolChain.cpp#L607), which [has this same darwin override](https://github.com/llvm/llvm-project/blob/bbc0f99f3bc96f1db16f649fc21dd18e5b0918f6/clang/lib/Driver/ToolChain.cpp#L577). We can't call that without instantiating a `clang::ToolChain()` though, and [given all the ceremony you added to do so elsewhere](https://github.com/apple/swift/blob/eee9331d97c348c87e8ae52901c7f0c3309d9a39/lib/ClangImporter/ClangIncludePaths.cpp#L440), that would end up with more lines.

@compnerd, I don't think that clang method is worth invoking, let me know if you disagree.

finagolfin commented:
@drodriguez, the Android CI [no longer builds the stdlib since last week](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android-arm64/4357/console), because #66334 calls `process_vm_readv()` that [was added in Android API 23](https://android.googlesource.com/platform/bionic/+/d633600d1c32eb00a51a521168eb336291847bb0/libc/include/sys/uio.h#145):
```
/home/ubuntu/jenkins/workspace/oss-swift-RA-linux-ubuntu-16.04-android-arm64/swift/stdlib/public/runtime/CrashHandlerLinux.cpp:665:12: error: use of undeclared identifier 'process_vm_readv'
    return process_vm_readv(memserver_pid, &local, 1, &remote, 1, 0);
           ^
1 error generated.
```
If I cross-compile the stdlib to API 24, which is what I usually do, I don't see this problem. I think you'll want to update the Android CI to API 23 to remedy this.

Once that's fixed, this pull should get the CI green again, and I can add some doc updates here about the new minimum Android API.

finagolfin commented:
Rebased and dropped the `noundef` change for now, as that test was recently changed in #67092 and I'm not sure if it still fails.

@drodriguez, do you want to increase the Android API level tested on the CI to get it building again?

finagolfin commented:
Rebased, fixed some more tests failing on Android, and updated the `noundef` change for that test now using opaque pointers.

finagolfin commented on test/Interop/Cxx/class/constructors-copy-irgen-android.swift:
```diff
@@ -9,13 +9,9 @@ import Constructors
-// ITANIUM_ARM: call void @_ZN30HasUserProvidedCopyConstructorC2ERKS_(%struct.HasUserProvidedCopyConstructor* [[ARG1_AS_STRUCT]], %struct.HasUserProvidedCopyConstructor* [[ARG2_AS_STRUCT]])
+// ITANIUM_ARM-SAME: (ptr {{.*}}[[ARG0:%.*]], ptr {{.*}}[[ARG1:%.*]], ptr {{.*}}[[ARG2:%.*]])
+// ITANIUM_ARM: call void @_ZN30HasUserProvidedCopyConstructorC2ERKS_(ptr [[ARG0]], ptr [[ARG2]])
+// ITANIUM_ARM: call void @_ZN30HasUserProvidedCopyConstructorC2ERKS_(ptr [[ARG1]], ptr [[ARG2]])
```
These constructor tests started failing recently and had to be updated for Android similarly to #67092.

finagolfin commented on test/Reflection/conformance_descriptors.swift:
```diff
@@ -13,7 +13,7 @@
 // CHECK: CONFORMANCES:
 // CHECK: =============
-// CHECK-DAG: 16ConformanceCheck10SomeStructV6${{[0-9a-f]*}}yXZ0c6NestedD06${{[0-9a-f]*}}LLV (ConformanceCheck.SomeStruct.(unknown context at ${{[0-9a-f]*}}).(SomeNestedStruct in ${{[0-9a-f]*}})) : ConformanceCheck.MyProto
+// CHECK-DAG: 16ConformanceCheck10SomeStructV{{[56]}}${{[0-9a-f]*}}yXZ0c6NestedD0{{[56]}}${{[0-9a-f]*}}LLV (ConformanceCheck.SomeStruct.(unknown context at ${{[0-9a-f]*}}).(SomeNestedStruct in ${{[0-9a-f]*}})) : ConformanceCheck.MyProto
```
I don't know why this started failing all of a sudden, as the length of this symbol on Android was 6 and it passed before, but lately it is 5, so this change gets it passing again.

finagolfin replied:
@artemcm, this test is disabled on all other arm64 platforms, is this the reason? Should I just disable it for Android AArch64 too?

finagolfin commented:
Alright, #67478 got the stdlib building again and turned up some more failing tests in the compiler validation suite. I will look into those and update this pull.

finagolfin commented on test/IRGen/debug_fragment_merge.sil:
```diff
@@ -3,7 +3,7 @@
 // fragment is 32 bits long, which changes the number of checks as well as the
 // arithmethic on the bounds of each fragment.
-// UNSUPPORTED: OS=watchos
+// UNSUPPORTED: OS=watchos, OS=linux-androideabi
```
These two debug tests fail on 32-bit arches, as the comment says, so disable both for Android armv7 too.

compnerd replied:
I assume that this "works" but only because for ARM64 the environment is not EABI, so you end up with the OS of `linux-android` and that would not match?

finagolfin replied:
Yes, `linux-androideabi` only matches armv7, so this test is only disabled for that 32-bit arch.

finagolfin replied:
I've removed these `debug_fragment` test changes, as more general requirements are being put in by others in #67699.

finagolfin commented on test/Interop/Cxx/templates/class-template-non-type-parameter-irgen.swift:
```diff
@@ -5,5 +5,5 @@ import ClassTemplateNonTypeParameter
-// CHECK: @"${{s4main1pSo0034MagicArrayInt32_UInt_2_zoAFhhiEngcVvp|s4main1pSo0036MagicArrayInt32_UInt64_2_JsAEiFiuomcVvp}}"
-// CHECK: @"${{s4main1tSo0034MagicArrayInt32_UInt_3_zoAFhhiEngcVvp|s4main1tSo0036MagicArrayInt32_UInt64_3_JsAEiFiuomcVvp}}"
+// CHECK: @"${{s4main1pSo0034MagicArrayInt32_UInt_2_zoAFhhiEngcVvp|s4main1pSo0036MagicArrayInt32_UInt(64|32)_2_JsAEiFiuomcVvp}}"
+// CHECK: @"${{s4main1tSo0034MagicArrayInt32_UInt_3_zoAFhhiEngcVvp|s4main1tSo0036MagicArrayInt32_UInt(64|32)_3_JsAEiFiuomcVvp}}"
```
This new test fails on 32-bit arches like Android armv7, but these two changes got it passing there too.

finagolfin replied:
@egorzhdan, you added this new test recently, please review this change to get it working on 32-bit platforms.

finagolfin commented on test/Interpreter/llvm_link_time_opt.swift:
```diff
@@ -1,7 +1,7 @@
 // static library is not well supported yet on Windows
 
-// REQUIRES: lld_lto
+// XFAIL: OS=linux-android, OS=linux-androideabi
```
This started failing recently on the community Android CI, as the LLVM 14 toolchain in the Android LTS NDK 25c doesn't support opaque pointers much:
```
error: Opaque pointers are only supported in -opaque-pointers mode (Producer: 'LLVM13.0.0git' Reader: 'LLVM 14.0.6git')
```
That flag wasn't added till clang 15, llvm/llvm-project@d69e9f9d89783, so this won't work till the next NDK, which will include clang 17. It currently works natively in the Termux app, where I use LLVM 16.

finagolfin commented on test/lit.cfg:
```diff
@@ -501,8 +501,6 @@ backtrace_on_crash = lit_config.params.get('backtrace_on_crash', None)
 if backtrace_on_crash is not None:
     config.environment['SWIFT_BACKTRACE'] = 'enable=on'
 
-config.available_features.add('lld_lto')
```
This was used to avoid issues on Android by @kateinoigakukun in #32430, but I enabled it for Android too when it switched to lls

finagolfin commented:
@egorzhdan, it's been a couple months since the CI was run on this, would you rerun?

egorzhdan commented:
@swift-ci please smoke test

egorzhdan commented:
@swift-ci please test

finagolfin commented:
@compnerd, @drodriguez appears to be away, would you review?

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1459,9 +1459,11 @@ void ToolChain::getClangLibraryPath(const ArgList &Args,
-                                         : getPlatformNameForTriple(T));
+  auto platformName =
+      T.isOSDarwin() ? "darwin" : getPlatformNameForTriple(Triple);
+  if (platformName == "android")
```
Would you mind using `T.isOSAndroid()` instead for this?

finagolfin replied:
Sure, `T.isAndroid()` is equivalent, fine by me, will make that change.

finagolfin replied:
@compnerd, made this change, let me know if it's okay in latest commit.

finagolfin commented:
@drodriguez, I see you're active again lately, any feedback on this pull? Just needs one last CI run and can be merged, should get the Android CI green again.

finagolfin commented:
Rebased and added a fix for a newly enabled C++ Interop test that fails on Android.

finagolfin commented on test/Interop/Cxx/stdlib/use-cxxstdlib-types-in-module-interface.swift:
```diff
@@ -40,7 +40,7 @@ public:
-// CHECK: func getFRTItems() -> std{{\.__1\.|\.}}vector<FRTType, allocator<FRTType>>
-// CHECK: func getPODPtrItems() -> std{{\.__1\.|\.}}vector<UnsafePointer<SimplePOD>, allocator<UnsafePointer<SimplePOD>>>
-// CHECK: func getMutPODPtrItems() -> std{{\.__1\.|\.}}vector<UnsafeMutablePointer<SimplePOD>, allocator<UnsafeMutablePointer<SimplePOD>>>
+// CHECK: func getPODItems() -> std{{\.__(ndk)?1\.|\.}}vector<SimplePOD, allocator<SimplePOD>>
```
Otherwise fails on Android with this error:
```
/home/ubuntu/jenkins/workspace/oss-swift-RA-linux-ubuntu-16.04-android-arm64/swift/test/Interop/Cxx/stdlib/use-cxxstdlib-types-in-module-interface.swift:43:11: error: CHECK: expected string not found in input
// CHECK: func getPODItems() -> std{{\.__1\.|\.}}vector<SimplePOD, allocator<SimplePOD>>
          ^
<stdin>:1:1: note: scanning from here

^
<stdin>:16:2: note: possible intended match here
 func getPODItems() -> std.__ndk1.vector<SimplePOD, allocator<SimplePOD>>
 ^
```

finagolfin commented:
@bnbarham, please run the CI on this.

bnbarham commented:
@swift-ci please test

finagolfin commented:
@compnerd, please merge if you're okay with the two changes I made since you approved, which are called out above.

compnerd reviewed (approved):
Trying to simplify `getClangLibraryPath` might be a nice follow up change.
